tools/generate-changelog.py

```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python3
+
+import datetime
+from pathlib import Path
+import re
+import subprocess
+
+import click
+
+LINK_RE = re.compile(r"\[#(\d+)\]")
+ISSUE_URL = "https://github.com/trezor/trezor-firmware/issues/{issue}"
+
+VERSION_HEADER_RE = re.compile(r"## \[([.0-9]+)\]")
+DIFF_LINK = "[{new}]: https://github.com/trezor/trezor-firmware/compare/{tag_prefix}{old}...{tag_prefix}{new}\n"
+
+
+def linkify_changelog(changelog_file, only_check=False):
+    links = {}
+    orig_links = {}
+    result_lines = []
+
+    with open(changelog_file, "r+") as changelog:
+        for line in changelog:
+            m = LINK_RE.match(line)
+            if m:  # line *starts with* issue identifier
+                # keep existing links as-is
+                orig_links[int(m[1])] = line.replace(m[0] + ": ", "").strip()
+            else:
+                for issue in LINK_RE.findall(line):
+                    links[int(issue)] = ISSUE_URL.format(issue=issue)
+                result_lines.append(line)
+
+        if only_check:
+            missing_links = set(links.keys()) - set(orig_links.keys())
+            if missing_links:
+                click.echo(f"missing links: {missing_links}")
+                return False
+            else:
+                return True
+
+        links.update(orig_links)
+
+        changelog.seek(0)
+        changelog.truncate(0)
+        for line in result_lines:
+            changelog.write(line)
+        for marker, url in sorted(links.items()):
+            changelog.write(f"[#{marker}]: {url}\n")
+
+    return True
+
+
+def linkify_gh_diff(changelog_file, tag_prefix):
+    linkified = False
+    versions = []
+    result_lines = []
+
+    with open(changelog_file, "r+") as changelog:
+        for line in changelog:
+            m = VERSION_HEADER_RE.match(line)
+            if m:
+                versions.append(m[1])
+            result_lines.append(line)
+
+        changelog.seek(0)
+        changelog.truncate(0)
+        for line in result_lines:
+            changelog.write(line)
+            if not linkified and VERSION_HEADER_RE.match(line):
+                changelog.write(
+                    DIFF_LINK.format(
+                        tag_prefix=tag_prefix, new=versions[0], old=versions[1]
+                    )
+                )
+                linkified = True
+
+
+def current_date(project):
+    parts = project.parts
+    today = datetime.datetime.now()
+
+    if parts[-2:] == ("legacy", "bootloader"):
+        return today.strftime("%B %Y")
+    elif parts[-1] == "python":
+        return today.strftime("%Y-%m-%d")
+    else:
+        daysuffix = {1: "st", 2: "nd", 3: "rd"}.get(today.day % 10, "th")
+        return today.strftime(f"%d{daysuffix} %B %Y")
+
+
+@click.command()
+@click.argument(
+    "project",
+    type=click.Path(exists=True, dir_okay=True, resolve_path=True),
```

Comment:
should probably be `file_ok=False` ?

Reply:
Fixed in 5b80e81f4887f153875e7e37d4d81490cf414c12.